content/book/chapter_1/Chapter_1.md

```diff
@@ -0,0 +1,213 @@
+
+### Running with a database
+
+Again, see [the project repo](todo:link) for instructions on how to run the app with a database. You will need to install the database we use (called MySQL) onto your computer, then pass details of how to connect to the database to the webserver when you start it.
+
+Once you have those things running together then you can create Todo items using `curl` or your the web page, and you can restart the webserver without losing your data. You can also restart the database - those really do write the data into files on the disk, so we will find the same data still exists after a restart. We might draw the application architecture like this:
+
+![webapp plus database](./webapp2.png)
+
+In this architecture, we can refer to the webserver as being _stateless_ (in the sense that it isn't responsible for any of the application's _state_). This has a number of advantages, in particular it lets us create multiple identical webservers all connecting to the same database. This might let us server more users, and we can be confident in not losing data if a webserver crashes or hardware fails. Having multiple copies of the database is also possible - and is another good reason why we should use a dedicated database.
+
+*Aside:* Most architecture diagrams are just boxes and lines, but databases are usually drawn as cylinders. Why? Because they store data on disks and early disks were cylindrical:
+
+![disk](./disk.png)
+
+This is why Oracle (a database company originally) has buildings of that shape in its HQ:
+
+![Oracle hq](./orahq.png)
+
+(You might recognise the buildings the HQ of Cyberdyne systems in Terminator Genesis, too)
+
+
+## Accessing the application
+
+### Not everyone is using your computer
+
+(so we can't use localhost forever)
+
+### Accessing from the internet
+
+Your computer is probably not directly connected to the internet, but connects through your router or similar at home:
+
+![home server architecture](./webapp3.png)
+
+Configure router to forward connections, then you can use external IP to access your application at `http://external-ip:8080/` but this isn't ideal for a number of reasons:
+
+  - the wifi connection in your house might fail
+  - your computer might get turned off by a cat
+  - your ISP might decide to issue your router with a different IP address
+  - your house might have a power cut
+  - your ISP might not not let you have a very high upload speed
+  
+etc
+
+TODO: Mention DNS here
+
+To mitigate this in 2008, you would have rented some space - about the size of a pizza box - in a datacenter. Then drive there with a small computer and leave it there.:
+
+![basic DC arch](./webapp4.png)
+
+Todo: pictures and stats about DCs
+
+Problems which you still have:
```

Comment:
Add a bit more info here on what the issues are, but might not necessarily want to go into too much detail about this in the lecture.